### PR TITLE
initial pass at install | lint | format | it

### PR DESCRIPTION
**Pull request by DJRickyB**
With this commit we add barebones functionality for managing rally-tracks as a single project, with capabilities pre-requisite for Continuous Integration.  Specifically, we add a Makefile, with targets and configurations for:
* install
* lint
* format
* it

Note that the code does not currently "pass" this lightweight configuration, and changes meant to re-format or clean up the code will happen in a later PR.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	@. $(VENV_ACTIVATE_FILE); isort **
+
+docs: check-venv
+	@. $(VENV_ACTIVATE_FILE); cd docs && $(MAKE) html
+
+serve-docs: check-venv
+	@. $(VENV_ACTIVATE_FILE); cd docs && $(MAKE) serve
+
+test: check-venv
+	. $(VENV_ACTIVATE_FILE); pytest tests/
```
Comment (b-deam) [inline Makefile:76]: Do we also want to add pytest to `requirements.txt`?

**Review by b-deam — COMMENTED** [review]
I gave it a quick test and it works (except for `pytest` missing).

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+export TRACK_DIRS = $(shell find -E . -regex ".*/track\.json"|sed -r 's|/[^/]+$$||' | sort -u)
+
+prereq:
+	pyenv install --skip-existing $(PY38)
+	pyenv local $(PY38)
+	@# Ensure all Python versions are registered for this project
+	@ jq -r '.python_versions | [.[] | tostring] | join("\n")' .ci/variables.json > .python-version
+	-@ printf $(PYENV_PREREQ_HELP)
+
+venv-create:
```
Comment (ebadyano) [inline Makefile:42]: Do we want venv-create to be executed as part of make install, similarly to rally Makefile?

File: .pylintrc [hunk 3]
```diff
@@ -0,0 +1,588 @@
+[MASTER]
```
Comment (DJRickyB) [inline .pylintrc:1]: this file is the default generated from `pylint --generate-rcfile >.pylintrc`.  no special rules here as opposed to the Rally one.  As noted in the PR description, this repo actually fails linting currently so modifications to the ruleset as well as code cleanup will happen in a follow-up PR.

**Comment by DJRickyB** [comment]
This is re-ready for re-review :)

New approach is lighterweight and does not dictate its own separately-maintained Python or library versions for the (eventual) integration testing.  Please let me know what you think of the submodule thing here

**Review by pquentin — CHANGES_REQUESTED** [review]
I sympathize with your willingness to make progress, any progress, as the current situation is dire, but I'm worried that this goes into the wrong direction.

Using a git submodule is a hack and make(1) just feels like the wrong tool here, and we keep running into its limitations. If only we had a good cross-platform programming language well-suited to writing scripts, and a good way to import code from other repositories!

I'm joking: we do have that. We could install Rally with pip (even from git if we wanted too), add any shared code in esrally/testing.py and use nox to run lint and tests, dropping tox. I'm not sure how we should handle the virtual environment creation yet, as it's outside of nox's scope. There are many concerns here, which is why we should probably resume the offline discussion we had around testing this repository. It seems to me that we now have enough information about CI limitations to do that. cc @michaelbaamonde

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+init-venv: checkout-rally
+	@if [[ ! -f $(VENV_ACTIVATE_FILE) ]]; then \
+	printf "Initializing virtual environment for Rally submodule...\n"; \
+    cd rally && make venv-create; \
+	fi
+
+# keep development dependencies in sync with Rally
+install: init-venv
+	cd rally && make install
```
[the remark this answers is not in the document]
Comment (DJRickyB) [reply Makefile:45]: Philosophically, this is all CI machinery, for use in an ephemeral environment such as a Jenkins or GH Actions worker. Doesn't mean there's not a better way to do this but I wanted to point that out from the intent

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+SHELL = /bin/bash
+# We assume an active virtualenv for development
+PYENV_REGEX = .pyenv/shims
+PY_BIN = python3
+# https://github.com/pypa/pip/issues/5599
+PIP_WRAPPER = $(PY_BIN) -m pip
+RALLY_SUBMODULE = rally
+VENV_ACTIVATE_FILE = $(RALLY_SUBMODULE)/.venv/bin/activate
+VENV_ACTIVATE = . $(VENV_ACTIVATE)
```
[the remark this answers is not in the document]
Comment (DJRickyB) [reply Makefile:26]: yep!

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+export TRACK_DIRS = $(shell find -E . -regex ".*/track\.json"|sed -r 's|/[^/]+$$||' | sort -u)
```
[the remark this answers is not in the document]
Comment (DJRickyB) [reply Makefile:27]: yeah... why did i do that?

**Comment by DJRickyB** [comment]
Thanks @pquentin! I added you here for a reason :)

> Using a git submodule is a hack and make(1) just feels like the wrong tool here, and we keep running into its limitations. If only we had a good cross-platform programming language well-suited to writing scripts, and a good way to import code from other repositories!

The submodule bit was to consider matching compatible rally-tracks branches with compatible Rally versions, using the ref mechanism in non-main branches to ensure compatibility where we expected it, bi-directionally.  I think this can probably be done just as well with Pip, as you mention, but as I wanted to also source venv creation and compatible Python/dev dependencies, the submodule quickly became an all-in-one solution.
 
I think the hack can live here, maybe even justifiably, with an established entry point which from the CI machine's perspective wouldn't change even as we improve or re-factor under-the-hood.  So my immediate task will be to write a top-level CI script, but I'd be curious about your thoughts on the above set of goals and intentions before I go any further with the implementation detail

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (pquentin) [reply Makefile:45]: I see what you mean. But we'll also use this machinery, so it makes sense to optimize for us. If things are barely usable in .ci/build.sh, I don't really mind, but in tools we use such as the Makefile I think good UX is important.

**Comment by pquentin** [comment]
The arguments for the submodule are solid, thanks.

@michaelbaamonde Should we close this for now?

**Pull request closed without merging**